### ISSUE TITLE
Sorting `citation-number` expressed as “data” versus “display”?

### ISSUE THREAD

**Issue by Jason-Abbott** [citation-style-language/test-suite#62]
(My purpose is just to implement the clearest, most concise code that passes the tests. I’m happy to forego a conceptual treatise if you have a dozen lines of pseudo-code that do the trick.) 🙂

A variety of tests (`sort_BibliographyCitationNumber*`, `sort_CitationNumber*`) yield something like
```
[4] Xxxx (fourth-cited)
[3] Zzzz (third-cited)
[2] Bbbb (second-cited)
[1] Aaaa (first-cited)
```
for a 
```xml
<text variable="citation-number" prefix="[" suffix="] "/>
<text variable="title" />
```
layout sorted with
```xml
<key variable="citation-number" sort="descending" />
``` 
Since the [specification](https://docs.citationstyles.org/en/stable/specification.html#number-variables) defines `citation-number` as the “index (starting at 1) of the cited reference in the bibliography,” this is perplexing. How can the _first_ “reference in the bibliography” be given a number indicating it’s the _fourth_ “reference in the bibliography”?

@fbennett addressed this in an [Issue 1 comment](https://github.com/citation-style-language/test-suite/issues/1#issuecomment-282447348), suggesting a distinction between `citation-number` expressed as “item data” versus a “display variable,” where the latter is the “preexisting sort order” and the former, I guess, the _current_ order.

Since the Swift processor I’m working on performs sorting before rendering (of course), and sorting immediately updates the indexes of references, my rendered `citation-number` is always what’s _current_. Maybe I need a reference property that tracks its _previous_ `citation-number` for rendering but that doesn’t seem quite right either.

[ More specific questions coming … ]

**Comment by Jason-Abbott** [comment]
I will start with something very basic: why does a secondary `citation-number` sort do _anything_ in tests ([simplest example](https://github.com/citation-style-language/test-suite/blob/2bea9caf5a3d04b3e68d4ecc8028c472a4499728/processor-tests/humans/sort_BibliographyCitationNumberDescendingViaCompositeMacro.txt)) where all items were sorted by the primary key value?

```xml
<key variable="title" sort="ascending"/>
<key variable="citation-number" sort="descending"/>
```

The specification says, “a secondary sort, using the second sort key, is applied to items sharing the first sort key value.”

In the linked test and the example above, no items share “the first sort key value.” Their titles are all different.

**Comment by adam3smith** [comment]
I believe [this discussion on the Zotero forums](https://forums.zotero.org/discussion/71443/reverse-number-bibliography) is pertinent, but personally I also find the `citation-number` behavior here perplexing and at a minimum out of spec, arguably counter to it. As a practical matter, I'm pretty sure the above case doesn't come up a single time in the existing styles, which have just two basic sort behaviors related to `citation-number`: The most common numbering by order of appearance (i.e. sorting by citation-number in `citation` and with no explicit sort in bibliography) (e.g. `vancouver.csl`) and sorting by citation-number in text and alphabetical sorting in the bibliography, which means citation numbers appear out of numerical order in the document (e.g., `american-medical-association-alphabetical.csl`), so if you decide to mark this test as won't-pass in citeproc-swift, performance on all existing styles will remain identical.

**Comment by Jason-Abbott** [comment]
Setting aside _why_ a secondary `citation-number` sort applies, I’m also perplexed by _how_ it applies.

The [specification](https://docs.citationstyles.org/en/stable/specification.html#sorting) says that `<sort/>` affects the order of “cites within citations, and bibliographic entries within the bibliography.”

But in the examples noted above, the `citation-number` sort doesn’t do that.  The items retain the order they had before the `citation-number` sort. The only thing apparently sorted is the value of the `citation-number` itself.

I regret if this sounds critical. I fully understand how complex software evolves over time. 🙂 I’m totally happily to match citeproc-js behavior as soon I can distill it to a minimum number of logical statements.

**Comment by Jason-Abbott** [comment]
It looks like it began here (@fbennett [2011 Zotero comment](https://forums.zotero.org/discussion/comment/99175/#Comment_99175)):

> The current processor cannot do this: when sort is performed on the citation-number variable in descending order, it instead reverses the sequence of items, and then numbers them in ascending order, starting with 1. This seems to make little sense, so I have introduced a fix that will instead leave the citations in document order, and number them in descending order (as you require).

I’ll find that change set. I don’t imagine the code is too hard (famous last words, especially with CSL) — just some exceptional handling where `citation-number` is used. At the least, I’ll update the documentation in the relevant tests and comment here for future reference.

**Comment by Jason-Abbott** [comment]
It [looks like](https://github.com/Juris-M/citeproc-js/blob/73bc1b44bc7d54d0bfec4e070fd27f5efe024ff9/citeproc.js#LL11155C21-L11155C21) citeproc-js has completely separate sort handling when the key is `citation-number`. It’s not used for normal sorting at all.

```js
if (variable === "citation-number") {
  func = function(state, Item) {
    if (state.tmp.area === "bibliography_sort") {
      if (this.strings.sort_direction === CSL.DESCENDING) {
        state.bibliography_sort.opt.citation_number_sort_direction = CSL.DESCENDING;
      } else {
        state.bibliography_sort.opt.citation_number_sort_direction = CSL.ASCENDING;
[ … ]
```

I _think_ that means I can just ignore
```xml
<key variable="citation-number" sort="descending" />
```
for actual sorting purposes and instead use it to set a flag in state somewhere indicating that `citation-number` values should be inverted.

**Comment by Jason-Abbott** [comment]
[Also instructive](https://github.com/Juris-M/citeproc-js/blob/73bc1b44bc7d54d0bfec4e070fd27f5efe024ff9/src/state.js#L527)
```js
CSL.Engine.BibliographySort = function () {
    this.tokens = [];
    this.opt = {};
    this.opt.sort_directions = [];
    this.opt.topdecor = [];
    // Holds the final citation-number sort direction, for use
    // in applying numbers in cs:citation and cs:bibliography.
    // Value is exclusively controlled by cs:key in bibliography_sort
    this.opt.citation_number_sort_direction = CSL.ASCENDING;
    this.opt.citation_number_secondary = false;
    this.tmp = {};
    this.keys = [];
    this.root = "bibliography";
};
```
I think I can handle this at the XML parse phase — convert any `<sort key="citation-number"/>` to a state value and remove it so it’s not even seen by the renderer.

**Comment by Jason-Abbott** [comment]
And [here](https://github.com/Juris-M/citeproc-js/blob/73bc1b44bc7d54d0bfec4e070fd27f5efe024ff9/citeproc.js#L23365), for completeness, is where citeproc-js calculates reversed `citation-number`s based on the state flag:

```js
CSL.Registry.prototype.renumber = function () {
    var len, pos, item;
    // 19. Reset citation numbers on list items
    if (this.state.bibliography_sort.opt.citation_number_sort_direction === CSL.DESCENDING) {
        this.state.bibliography_sort.tmp.citation_number_map = {};
    }
    len = this.reflist.length;
    for (pos = 0; pos < len; pos += 1) {
        item = this.reflist[pos];
        // save the overhead of rerenderings if citation-number is not used in the style.
        item.seq = (pos + 1);
        if (this.state.bibliography_sort.opt.citation_number_sort_direction === CSL.DESCENDING) {
            this.state.bibliography_sort.tmp.citation_number_map[item.seq] = (this.reflist.length - item.seq + 1);
        }
        // update_mode is set to CSL.NUMERIC if citation-number is rendered in citations.
        if (this.state.opt.update_mode === CSL.NUMERIC && item.seq != this.oldseq[item.id]) {
            this.state.tmp.taintedItemIDs[item.id] = true;
        }
        if (item.seq != this.oldseq[item.id]) {
            this.return_data.bibchange = true;
        }
    }
};
```
Notably `this.reflist.length - item.seq + 1`

**Comment by fbennett** [comment]
(Thank you for being sympathetic about the complexity of it all. It means a
lot.)

**Comment by Jason-Abbott** [comment]
It appears there may be another caveat: a `<key variable="citation-number" sort="descending"/>` in primary position _does_ change item order (e.g. [this test](https://github.com/citation-style-language/test-suite/blob/2bea9caf5a3d04b3e68d4ecc8028c472a4499728/processor-tests/humans/sort_CitationNumberPrimaryDescendingViaVariableBibliography.txt)) but in secondary or later positions it only causes the numbers to display in reverse order without changing item order — another thing to check.

I feel like I’m getting close to something without too many lines of code that will pass these tests. 🤞 I’ll share the relevant portions of Swift code if I get there.

**Comment by Jason-Abbott** [comment]
> (Thank you for being sympathetic about the complexity of it all. It means a lot.)

@fbennett I was astonished and daunted by all of your commits as I scrolled back to find the first to support reversed `citation-number`s 😬. I’ve written code a lot of years but haven't committed that much to a single project. It’s impressive.

**Comment by Jason-Abbott** [comment]
I think I have all the tests passing with minimal code changes. The logic I applied was

**Decoding Phase**
- If a sort `<key variable="citation-number"/>` is present then update a state variable `displayReversedCitationNumbers = key.sort == "descending"`
    - If that _is not_ the first `<key/>` then remove it so it doesn’t affect the sorting or rendering phases
- If a sort `<key macro="[some-name]"/>` is present and the macro _only_ contains `<text variable="citation-number"/>` then update the same state variable, `displayReversedCitationNumbers = key.sort == "descending”`
    - If that _is_ the first `<key/>` element then remove it and replace it with `<key variable="citation-number"/>`
    - If it _is not_ the first `<key/> then remove it so it doesn’t affect the sorting or rendering phases

With that done there are no more
```xml
<macro name="citation-number">
  <text variable="citation-number"/>
</macro>
```
and no more 
```xml
<key variable="citation-number"/>
```
that do magical things. The `displayReversedCitationNumbers` state value has what’s needed.

**Sorting Phase**
- Here I just added a `reversedPosition` property to the reference object to go along with the `position` property I was already updating after sort. Obviously it could be calculated on-demand but but 🤷‍♂️ 
- If citations are sorted by `citation-number` then reverse the order if `displayReversedCitationNumbers == true` (one line change)

**Rendering Phase**
- Use `reversedPosition` for a rendered `citation-number` if `displayReversedCitationNumbers == true` (a couple lines changed)

That’s it. I won’t be at all surprised if I run into scenarios needing adjustments but this passes all `sort_BibliographyCitationNumber`* and `sort_CitationNumber`* tests, and hundreds more.
